[C++ API] Remove virtual forward and implement Sequential based on Any(Module)

goldsborough opened the pull request:
This PR removes the virtual `forward()` method from the base module class, thus freeing submodules from any constraints w.r.t. the signature of their `forward()` method. This will enable the majority of modules to give their `forward()` a signature that best matches their module.

Of course, a virtual `forward()` has its merits: one can call a module's `forward()` when one only has a pointer to a `Module` (the base class). In our API, this is useful in one and only one case: `Sequential`. To work around the fact that we have no `forward()` to call on a `Module*`, we employ type erasure techniques borrowed from `std::any`. More precisely:

1. I've created an `Any` abstraction that models `std::any` (C++17) and allows type erasure of a single plain value.
2. I've then created an `AnyModule` class, that:
    1. Uses the same type erasure technique to erase the concrete type of a module,
    2. Erases but stores the parameter types of the concrete module's `forward()`, and casts `Any`s passed at runtime to these stored types. 

This leads to the following API:

```
Sequential model(Linear(2, 3), Linear(3, 4), Linear(4, 5));
model.forward<Variable>(variable);
```

Here, `forward` accepts any arguments and forwards them to the first module. The types are checked at runtime. Because the return type of the last module passed to `Sequential` is also erased, `forward` is templatized over the return type  of the last module. This is somewhat unfortunate, but is the only intrusion to the user API besides the dynamic type checking. It may be a good idea to let the return type default to `Variable`, but I fear lots of users will run into exceptions when they don't know this and return something else instead, so maybe being explicit is more helpful.

I've written lots of tests for `Any`, `AnyModule` and `Sequential`, that also give you an idea of what the API looks like. Note that:

1. `Sequential` and `AnyModule` are part of the public API,
2. `Any` would ideally not be public, but has to be because it's returned by `Sequential`. I limit this fact by making `Any`'s constructor private and making construction possible only via a friend function in the `detail` namespace, which should hopefully deter users from using this for their own code.

I have also removed all existing containers from autogradpp.

Note: even though there is a shit ton of template programming going on, I believe the final `Sequential` container hides most of this complexity. Please suggest where more `static_assert`s or other things that can prevent template error spew could be placed.

__Bonus__: if you have any suggestions on how to simplify the code, especially the templaty parts, please suggest away!

@apaszke @zdevito @ebetica @jgehring 
@smessmer please take a look at the template metaprogramming parts.

### Future PRs

1. Change the core modules to have more sensible `forward()`s instead of just `variable_list`
2. Right now, all except the first module in the `Sequential` have to take a single argument. It would be nice if we could implement something so that when you return an `std::tuple` or some other type, we can automatically destructure it and thus allow multiple arguments in modules other than the first. Note that you can already take a `std::tuple` as the first argument, you'll just have to destructure it yourself.

goldsborough committed:
Remove virtual forward

ezyang commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -0,0 +1,215 @@
+  static_assert(
+      !std::is_lvalue_reference<Head>::value ||
+          std::is_const<typename std::remove_reference<Head>::type>::value,
+      "Modules stored inside AnyModule must not take (lvalue) references. Use pointers instead.");
```
Why not?

This error feels like it would be totally meaningless to me if I was not very well-versed the vagaries of C++ references.

goldsborough replied:
Because in this mechanism we are copying/movying parameters into `Any` objects that live somwhere on the stack, so it would be inherently unsafe to allow users to take lvalue refs to these values, especially since the values are different from the ones they pass into the function call! Should I simplify the error message and add a comment on top?

ezyang commented on test/cpp/api/any.cpp:
```diff
@@ -0,0 +1,200 @@
+    SECTION("int") {
+      auto any = make_any(5);
+      // const and non-const types have the same typeid()
+      REQUIRE(any.try_get<int>() != nullptr);
```
I guess, if you wanted to test that it actually returned 5, you'd have to say `*any.try_get<int>() == 5`, which would then promptly segfault if you had got it wrong?

ezyang commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+/// type erased object, whose concrete value can be retrieved at runtime by
+/// checking if the `typeid()` of a requested type matches the `typeid()` of the
+/// object stored. It is simplified in that it does not handle copying, as we do
+/// not require it for our use cases. Moves are sufficient.
```
I'm not sure how I feel about the fact that we're making our own implementation of `std::any`. A lot of details to get right when doing this.

goldsborough replied:
I looked quite closely at implementations of `any` in boost and libc++: https://github.com/llvm-mirror/libcxx/blob/master/include/any

I think the benefit of rolling our own is that we can lose a lot of the complexity of the implementation because we don't need many of the features, such as the ability to copy/clone any values. I also didn't see any nice, standard conformant open source implementation of `std::any`

goldsborough replied:
Also, if we want to provide a no RTTI path one day, we'll have more freedom to add it to our own implementation. But no doubt that rolling our own is always sketchy

smessmer replied:
Can we copy over the boost::any implementation? Your argument holds true as long as the Any type is only used here, but I'll probably have a use case in c10 too, which might need a slightly different feature set. Having a more feature-complete implementation closer to the standard might do us some good in the future. Also because its behavior is more well-known and predictable than hand-rolling our own.

ezyang commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+ public:
+  /// Default construction is disallowed. Thus our invariant: an `Any` contains
+  /// an object at all times, from its construction, to its destruction.
+  Any() = delete;
```
This is another difference from C++17 `std::any`, which has a default constructor and an "empty" state.

goldsborough replied:
Yes, but this is where I like the fact that we can simplify our implementation of any

ezyang commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+    /// A template because T&& would not be universal reference here.
+    template <typename U>
+    explicit Holder(U&& value_) noexcept
+        : Placeholder(typeid(T)), value(std::forward<U>(value_)) {}
```
RTTI! Good to note this out in a comment.

ezyang commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+  template <typename T>
+  explicit Any(T&& value)
+      : content_(
+            torch::make_unique<Holder<decay_t<T>>>(std::forward<T>(value))) {}
```
What happens if I try to `make_any<int&>`?

ezyang commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+  /// parameter matches the type of the object stored, and returns a null
+  /// pointer otherwise.
+  template <typename T>
+  T* try_get() {
```
Confused about ownership. Does this return a non-owning pointer? Then the dereference in `get()` seems to cause a copy to happen. Is this intentional?

ezyang commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+    explicit Placeholder(const std::type_info& type_info_) noexcept
+        : type_info(type_info_) {}
+    virtual ~Placeholder() = default;
+    const std::type_info& type_info;
```
Worth noting that this is a reference because `type_info` is not copyable.

goldsborough replied:
And because `typeid()` returns this type

smessmer replied:
isn't there a `type_index` type as well? Or does that need RTTI?

ezyang commented on test/cpp/api/any.cpp:
```diff
@@ -0,0 +1,200 @@
+  SECTION("type_info is correct") {
+    SECTION("int") {
+      auto any = make_any(5);
+      REQUIRE(any.type_info().hash_code() == typeid(int).hash_code());
```
Can't you just do a direct `==` here? Why is the `hash_code` necessary

goldsborough replied:
> There is no guarantee that the same std::type_info instance will be referred to by all evaluations of the typeid expression on the same type, although std::type_info::hash_code of those type_info objects would be identical, as would be their std::type_index.

ezyang commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -0,0 +1,215 @@
+
+/// A class to store a type erasd module, whose `forward()` method can be
+/// invoked, with dynamic type checking.
+class AnyModule {
```
Do implicit conversions work when I invoke an `AnyModule`?

ezyang commented on test/cpp/api/any.cpp:
```diff
@@ -0,0 +1,200 @@
+    };
+    AnyModule any(M{});
+    REQUIRE_THROWS_WITH(
+        any.forward(5.0),
```
I guess this test suggests no implicit conversions happen. That's annoying.

goldsborough replied:
No! Sad. We could add manual stuff to do implicit conversions for primitive types? Should we?

ebetica replied:
I think implicit conversions would be cool. Another quality of life thing is to make a default `.get()` return an `at::Tensor`, since in a lot of cases you get 1-var input and outputs.

goldsborough commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -0,0 +1,172 @@
+  using Iterator = std::vector<std::shared_ptr<AnyModule>>::iterator;
+
+  /// Constructs the `Sequential` from a pack of modules. Each module can either
+  /// be a plain value (e.g. `Linear`) or a boxed value (e.g.
```
No if you pass an lvalue (i.e. not a temporary) into `push_back`, it will just copy it into a newly created boxed instance internally. Since `Variable`/`Tensor` is itself refcounted, this copy would lead to shared ownership of variables. If you pass a temporary or `std::move` a module before passing it in, the `Sequential` will take ownership.

I need to think a bit harder about what ownership semantics modules should have in general: value, reference, copy etc.

goldsborough commented on test/cpp/api/any.cpp:
```diff
@@ -0,0 +1,200 @@
+    SECTION("references") {
+      std::string s("hello");
+      const std::string& t = s;
+      auto any = make_any(t);
```
I was kind of regarding it as a black box here

smessmer commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+    static_assert(
+        !std::is_array<T>::value,
+        "Any stores decayed types, you must cast it to T* instead of T[]");
+    if (typeid(T).hash_code() == type_info().hash_code()) {
```
I don't think you can rely on `hash_code` being unique.

goldsborough replied:
So cppreference says:
> There is no guarantee that the same std::type_info instance will be referred to by all evaluations of the typeid expression on the same type, although std::type_info::hash_code of those type_info objects would be identical, as would be their std::type_index.

At the same time, libc++ just compares the values returned by `typeid()`: https://github.com/llvm-mirror/libcxx/blob/master/include/any#L166

`hash_code()` sounds safer to me?

smessmer replied:
Possibly clang gives this guarantee but the standard doesn't. We shouldn't rely on that either then.
`hash_code` is wrong the other way round. While comparing `typeid` directly might give you false negatives, `hash_code` might give you false positives. Can you use the `std::type_index` for comparison or does that need RTTI? If yes, how does boost::any solve it?

goldsborough replied:
`std::type_index` has the same guarantees as `hash_code`. The above quote is from cppreference, so it should be standard?

smessmer replied:
Quote from cppreference:

> No other guarantees are given: type_info objects referring to different types may have the same hash_code (although the standard recommends that implementations avoid this as much as possible), and hash_code for the same type can change between invocations of the same program.

smessmer commented on torch/csrc/api/include/torch/detail/any.h:
```diff
@@ -0,0 +1,138 @@
+
+template <typename Head, typename... Tail>
+void push_any(std::vector<Any>& vector, Head&& head, Tail&&... tail) {
+  vector.push_back(make_any(std::forward<Head>(head)));
```
This needs to move the any. An implementation with `emplace_back` might be able to avoid that.

goldsborough replied:
Actually the `Any` constructor is private

smessmer commented on test/cpp/api/any.cpp:
```diff
@@ -0,0 +1,200 @@
+    }
+    SECTION("const int") {
+      auto any = make_any(5);
+      REQUIRE(any.try_get<const int>() != nullptr);
```
I'm missing tests that have the wrong type and actually do return `nullptr`.

smessmer commented on torch/csrc/api/include/torch/nn/module.h:
```diff
@@ -148,4 +148,11 @@ std::shared_ptr<typename std::decay<Module>::type> make(Module&& module) {
+using is_module = std::is_base_of<Module, typename std::decay<M>::type>;
+
+template <typename M>
+using enable_if_module = typename std::enable_if<is_module<M>::value>::type;
```
Without the `_t` prefix, convention is to not go into the `::type` member, i.e.

    template <typename M>
    using enable_if_module = std::enable_if<is_module<M>::value>;
   
    template <typename M>
    using enable_if_module_t = typename enable_if_module<M>::type;

smessmer commented on torch/csrc/utils/variadic.h:
```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
+using disable_if_t = enable_if_t<!value, T>;
+
+template <typename T>
+using decay_t = typename std::decay<T>::type;
```
Can you only define these if not already present, and if present do `using enable_if_t = std::enable_if_t` instead?

smessmer commented on torch/csrc/utils/variadic.h:
```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
+template <bool... values>
+struct all_of : std::is_same<
+                    detail::pack<values..., true>,
+                    detail::pack<true, values...>> {};
```
cool implementation trick. TIL.

smessmer commented on torch/csrc/utils/variadic.h:
```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
+
+template <typename T, typename... Ts>
+using disable_for_t =
+    enable_if_all_of_t<(!std::is_same<T, decay_t<Ts>>::value)...>;
```
Let's add some (static_assert based) test cases for these type traits. To save compile time, probably rather in a test.cpp file and not in the header.

smessmer commented on torch/csrc/utils/variadic.h:
```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
+                    detail::pack<true, values...>> {};
+
+template <bool...>
+struct any_of_t;
```
Usually makes sense to define `any_of` first and then `any_of_t` based on it. Later in c++17, someone probably wants to add `any_of_v` based on the same implementation.

smessmer commented on torch/csrc/utils/variadic.h:
```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
+struct any_of_t<> : std::false_type {};
+
+template <bool head, bool... tail>
+struct any_of_t<head, tail...> {
```
`any_of` is basically `std::disjunction` and `all_of` is `std::conjunction`. Maybe makes sense to name them accordingly, and then also take the "official" implementation from [here](http://en.cppreference.com/w/cpp/types/conjunction) to make sure it behaves correctly and we didn't miss any corner cases.

Note: `std::conjunction` and `std::disjunction` have a slightly different interface because they use types instead of bool template arguments, but that actually makes it more versatile. You can still use it with bools together with `std::bool_constant` if you need to.

    template<bool... args> using any_of = std::disjunction<std::bool_constant<args>...>;

goldsborough replied:
`std::conjunction` and `std::disjunction` are C++17?

smessmer replied:
Yes, I meant it makes sense to implement our versions of `std::conjunction` and `std::disjunction` instead of implementing a similar-but-different `any_of` and `all_of`. You can take the implementation from cppreference.

smessmer commented on torch/csrc/utils/variadic.h:
```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
+using enable_if_all_of_t = enable_if_t<all_of<values...>::value>;
+
+template <typename T, typename... Ts>
+using disable_for_t =
```
I find this name a bit confusing. Maybe `disable_if_contains_t`? Also the auto-decaying is a bit hard to explain to the call site. Maybe keep that out of here and move it to the call site instead? Else, maybe call it `disable_if_contains_decayed_t`?

smessmer commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -0,0 +1,215 @@
+namespace torch {
+namespace detail {
+template <typename Head = void, typename... Tail>
+void check_not_lvalue_references() {
```
note: This can also be implemented using `std::conjunction` and `std::disjunction` (see my comments below)

smessmer commented on torch/csrc/utils/memory.h:
```diff
@@ -0,0 +1,38 @@
+}
+
+template <typename T>
+typename unique_type_for<T>::unbounded_array make_unique(size_t size) {
```
Can you add test cases for this file?

smessmer commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -0,0 +1,172 @@
+  /// module with the input of the next, in order of construction.
+  template <typename ReturnType, typename... ArgumentTypes>
+  ReturnType forward(ArgumentTypes&&... arguments) {
+    AT_CHECK(!is_empty(), "Cannot call forward() on an empty Sequential");
```
Why not? If it semantically makes sense (which I think it does), it might make sense to just return the input directly instead of erroring.

goldsborough replied:
What if there are multiple inputs?

smessmer replied:
Do I get it right that you're special casing the first op in the sequence to allow multiple inputs, whereas further ops can only take one input? If yes, then multiple inputs and no components is probably an error case.

If there's no special case, say intermediate components can have multiple outputs, then multiple inputs and no component would be the same as having the last component output multiple outputs.

I'm not 100% sure what the best behavior is, I'm trying to make it follow semantics and avoid edge cases that need to error out.

smessmer commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -0,0 +1,172 @@
+  template <typename ModuleType>
+  void push_back(std::shared_ptr<ModuleType> module_ptr) {
+    // Nesting Sequential doesn't work because `forward()`'s return type is
+    // templatized, so it'll give a nasty compiler error.
```
It wouldn't be too hard to make this possible, not sure if it's needed though.

smessmer commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -0,0 +1,172 @@
+/// The `Sequential` class reference semantics.
+class Sequential : public CloneableModule<Sequential> {
+ public:
+  using Iterator = std::vector<std::shared_ptr<AnyModule>>::iterator;
```
Why not `::iterator` (small case) as in the STL?

goldsborough replied:
All our types are capitalized

smessmer commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -0,0 +1,172 @@
+      typename... Modules,
+      typename = disable_for_t<Sequential, Modules...>>
+  explicit Sequential(Modules&&... modules) {
+    push_back(std::forward<Modules>(modules)...);
```
reserve before?

smessmer commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -0,0 +1,215 @@
+  /// Attempts to cast the underlying module to the given module type. Throws an
+  /// exception if the types do not match.
+  template <typename T, typename = enable_if_module<T>>
+  T& get() {
```
This is inconsistent to your `Any` implementation which returns by value. If you decide to change this here instead of the one in Any, you probably have to change the call site in Sequential so it doesn't suddenly copy.

goldsborough replied:
`Any` is an implementation detail and is used to store temporaries, while `AnyModule` is used to store modules permanently

smessmer replied:
This restriction on `Any` isn't obvious from the code. Can we make `Any` generic enough to handle temporaries and permanent values? If yes, then it could be reused here.

smessmer commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -0,0 +1,215 @@
+  /// The static type of the object we store in the `AnyModule`, which erases
+  /// the actual type, but allows us to call `forward()` on the underlying
+  /// module.
+  struct Placeholder {
```
I wouldn't implement this stuff twice. Have you considered using an `Any` member inside of `AnyModule` to solve this?

goldsborough replied:
I can see if refactoring out `Any`'s `Placeholder` makes the code shorter. Assuming we stick with our own `Any`

goldsborough commented:
I guess why not copy https://github.com/llvm-mirror/libcxx/blob/master/include/any?

Or https://www.boost.org/doc/libs/1_53_0/boost/any.hpp and remove the boost headers?

goldsborough commented:
Updates:
- Merged with latest changes from master
- Added tests for static type traits I added
- Added tests for make_unique
- Addressed various review nits

goldsborough commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -0,0 +1,172 @@
+
+  /// Adds a new (boxed) `Module` to the `Sequential` container.
+  template <typename ModuleType>
+  void push_back(std::shared_ptr<ModuleType> module_ptr) {
```
The `static_assert` makes it nice!

```
/home/psag/pytorch/pytorch/tools/cpp_build/libtorch/../../../torch/csrc/api/include/torch/nn/modules/sequential.h: In instantiation of ‘void torch::nn::Sequential::push_back(std::shared_ptr<_Tp1>) [with ModuleType = ____C_A_T_C_H____T_E_S_T____0()::X]’:
/home/psag/pytorch/pytorch/tools/cpp_build/libtorch/../../../torch/csrc/api/include/torch/nn/modules/sequential.h:94:5:   required from ‘void torch::nn::Sequential::push_back(M&&) [with M = ____C_A_T_C_H____T_E_S_T____0()::X]’
/home/psag/pytorch/pytorch/test/cpp/api/sequential.cpp:29:29:   required from here
/home/psag/pytorch/pytorch/tools/cpp_build/libtorch/../../../torch/csrc/api/include/torch/nn/modules/sequential.h:76:5: error: static assertion failed: Can only add objects derived from nn::Module to Sequential
     static_assert(
     ^
```

goldsborough commented:
I've added a way of giving users nice `static_assert`s when they try to add a module to `Sequential` that doesn't have a `forward()`:

```
struct X : public Module {};
sequential.push_back(X());
```

gives

```
/home/psag/pytorch/pytorch/tools/cpp_build/libtorch/../../../torch/csrc/api/include/torch/nn/modules/sequential.h:98:5:   required from ‘void torch::nn::Sequential::push_back(M&&) [with M = ____C_A_T_C_H____T_E_S_T____0()::X]’
/home/psag/pytorch/pytorch/test/cpp/api/sequential.cpp:29:29:   required from here
/home/psag/pytorch/pytorch/tools/cpp_build/libtorch/../../../torch/csrc/api/include/torch/nn/modules/sequential.h:80:5: error: static assertion failed: Can only add modules with a forward() method to Sequential
     static_assert(
```

goldsborough commented:
After discussions with @zdevito, I have turned the `Any` class, which was sort of general purpose, into a nested class of `AnyModule`. This reduces its scope and thus solves our discussion about how standardized the `Any` class needs to be.

Otherwise, I think this is ready now. @apaszke, @smessmer and @ezyang (after PTO) please re-review and sign off.

ebetica commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -0,0 +1,347 @@
+
+  /// Constructs an `AnyModule` from a `shared_ptr` to concrete module object.
+  template <typename ModuleType>
+  explicit AnyModule(std::shared_ptr<ModuleType> module);
```
Can we get this for operator= as well? Because in module construction, sometimes it's nice to use AnyModule in a Module, which means it's predeclared somewhere and I need to assign it to another module.

goldsborough replied:
yes I'll add it

goldsborough replied:
Actually this just works naturally in C++, since the right hand side of an `=` expression will construct a new `AnyModule`, which is then move-assigned:
```cpp
struct M : nn::Module {
      explicit M(int value_) : value(value_) {}
      int value;
      int forward(float x) {
        return x;
      }
    };
AnyModule any;
REQUIRE(any.is_empty());
any = AnyModule(M{5});
REQUIRE(!any.is_empty());
REQUIRE(any.get<M>().value == 5);
```

goldsborough replied:
But I've added an empty state

ebetica replied:
Sorry, I meant something like

```
any = M{5};
```

goldsborough commented:
![image](https://user-images.githubusercontent.com/6429851/40323995-f3e101e2-5ceb-11e8-8dec-f1525596b203.png)

goldsborough committed:
Rebase

goldsborough commented on test/cpp/api/util.h:
```diff
@@ -0,0 +1,45 @@
+  }
+};
+
+struct SigmoidLinear : nn::Module {
```
It's in `util.h` because it's used in two different test files (notice this is `test/...`).
They need to be placed in a `Sequential`, and I haven't created activation modules yet to compose them like in PyTorch

ebetica replied:
Ah I didn't see this, but I think you should use the Functional module to define these simple activation functions. I wanted to keep the nn:: namespace as clean a possible.

zdevito reviewed (approved):
(stamp)

goldsborough commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -0,0 +1,172 @@
+  /// module with the input of the next, in order of construction.
+  template <typename ReturnType, typename... ArgumentTypes>
+  ReturnType forward(ArgumentTypes&&... arguments) {
+    AT_CHECK(!is_empty(), "Cannot call forward() on an empty Sequential");
```
I figured this was tricky, because what value do you return if you get multiple input arguments :/
